### PR TITLE
Introduce optional uniform appearance of `Modal` #DS-1091

### PR DESCRIPTION
**Pull request by adamkudrna**
## Description

Snímek obrazovky 2023-12-19 v 14 39 39

Add `spirit-feature-modal-enable-uniform-dialog` feature class to enable uniform appearance of `Modal` across all breakpoints.

Current mobile design is then accessible using the `Modal--dockOnMobile` modifier class.

### Additional context

We need to make sure it works in all scenarios:

1. Current behavior must remain intact.
2. New appearance can be turned on via CSS class,
3. … or via Sass feature flag.

### Issue reference

https://jira.lmc.cz/browse/DS-1091

### CODE REVIEW

File: packages/web/src/scss/components/Modal/README.md [hunk 1]
```diff
@@ -352,6 +352,21 @@ Both trigger and close buttons use `data` attributes to open and close the Modal
 | `data-spirit-target`                  | `string` | —       | ✔        | Target selector                                       |
 | `data-spirit-toggle`                  | `string` | `modal` | ✕        | Iterable selector                                     |
 
+## Feature Flag: Uniform Appearance on All Breakpoints
+
+The uniform appearance of modal dialog on all breakpoints is disabled by default. To enable it, either set the
+`$modal-enable-uniform-dialog` feature flag to `true` or use the `spirit-modal-enable-uniform-dialog` CSS class on any
+parent of the dropdown.
```
Comment (crishpeen) [inline packages/web/src/scss/components/Modal/README.md:359]: dropdown? :)

File: packages/web/src/scss/components/Modal/_ModalDialog.scss [hunk 2]
```diff
@@ -4,10 +4,52 @@
 // 3. We need to set the box-sizing again because the parent element unsets styles using `all: unset`.
 
 @use 'sass:map';
+@use '../../settings/feature-flags';
 @use '../../tools/breakpoint';
 @use '../../tools/typography';
 @use 'theme';
 
+@mixin -uniform-modal-dialog() {
```
Comment (crishpeen) [inline packages/web/src/scss/components/Modal/_ModalDialog.scss:12]: I thought we kept all mixins and functions in component tools. 🤔

Comment (crishpeen) [reply packages/web/src/scss/components/Modal/_ModalDialog.scss:12]: Maybe they are here because they are local? 

The problem with local mixins is that we are not able to test them :/

File: packages/web/src/scss/components/Modal/_ModalDialog.scss [hunk 3]
```diff
@@ -46,14 +88,42 @@
...
     }
 }
 
-.ModalDialog--expandOnMobile {
-    @include breakpoint.down(map.get(theme.$breakpoints, tablet)) {
-        height: theme.$dialog-max-height;
+// @deprecated The "uniform" dialog variant is deprecated and will be removed in the next major release.
+// Migration: Remove the feature flag and make the uniform dialog the default.
+@if feature-flags.$modal-enable-uniform-dialog {
+    .ModalDialog--docked {
```
Comment (crishpeen) [inline packages/web/src/scss/components/Modal/_ModalDialog.scss:108]: I am not 100 % sure we shouldn't somehow mention in the modifier name that docked is the ModalDialog only on the mobile breakpoint.

**Review by crishpeen — COMMENTED** [review]
LGTM, I have just a few small thoughts.

File: packages/web/src/scss/components/Modal/README.md [hunk 4]
```diff
@@ -369,3 +384,5 @@ Or feel free to write controlling scripts yourself.
 [scroll-view]: https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web/src/scss/components/ScrollView/README.md
 [web-readme]: https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web/README.md
 [dictionary-alignment]: https://github.com/lmc-eu/spirit-design-system/blob/main/docs/DICTIONARIES.md#alignment
+[readme-deprecations]: https://github.com/lmc-eu/spirit-design-system/tree/main/packages/web/README.md#deprecations
```
Comment (pavelklibani) [inline packages/web/src/scss/components/Modal/README.md:387]: Not sure who it was, but someone told me that we are mixing blob/tree links.

File: packages/web/src/scss/components/Modal/index.html [hunk 5]
```diff
@@ -761,4 +761,172 @@ <h2 id="example_disabled_backdrop_click_title" class="ModalHeader__title">Modal
...
+      data-spirit-toggle="modal"
+      data-spirit-target="#example_docked"
+      aria-controls="example_docked"
+      aria-expanded="false"
+    >
+      Open Docked Modal (mobile only)
+    </button>
+
+    <!-- Modal: start -->
+    <dialog id="example_uniform" class="Modal" aria-labelledby="example_uniform_title">
```
Comment (pavelklibani) [inline packages/web/src/scss/components/Modal/index.html:792]: Should't be IDs in kebab-case as we agreed on dlouha temata?

**Review by pavelklibani — APPROVED** [review]
Just two questions, otherwise LGTM

File: packages/web/src/scss/components/Modal/_ModalDialog.scss (hunk 2 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/_ModalDialog.scss:12]: Yes, these are just temporary local mixins, not intended for reusing. I don't think that testing them would be any helpful… We will remove them in the next major version anyway.

File: packages/web/src/scss/components/Modal/_ModalDialog.scss (hunk 3 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/_ModalDialog.scss:108]: OK, let's go with `dockOnMobile` (similar to `expandOnMobile`).

File: packages/web/src/scss/components/Modal/README.md (hunk 4 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/README.md:387]: I checked [what's behind it](https://stackoverflow.com/questions/39400848/in-github-urls-what-is-the-difference-between-a-tree-and-a-blob) and both are correct, each for different use case:

- **blob** is a Git object for file,
- **tree** is for directory.

Which means you are right, `blob` should be here instead of `tree`. 👍🏻

File: packages/web/src/scss/components/Modal/index.html (hunk 5 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/index.html:792]: Indeed, they should. I would rather change them all once all Modal updates are done—a lot of rebasing can be expected down the road and the fewer changes are all over the place, the better.

File: packages/web/src/scss/components/Modal/_ModalDialog.scss (hunk 2 above)
Comment (literat) [reply packages/web/src/scss/components/Modal/_ModalDialog.scss:12]: Please, be careful. Not testing something because it will be removed in the future can be problematic. We plan the next major version at the start of the Q2. So this code will be here for a while. Just saying.

**Review by literat — APPROVED** [review]
I have written down a few comments, otherwise the appearance is looking good. Nice work 👍

File: packages/web/src/scss/components/Modal/_ModalDialog.scss (hunk 2 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/_ModalDialog.scss:12]: Well, until a few weeks ago, we didn't have tests in Sass at all… What I'm trying to say here is this code only puts aside a few lines of other code. Plus we can expect it will have a very short life, but that's the less important thing.

A test could only check if the mixin prints exactly what is inside it. It doesn't programmatically generate other code, nor there are any injected dependencies. Thus it behaves deterministic: as long as the theme variables remain the same, it's fine. If they don't, we have a much bigger problem anyway.

To sum up, I get what you mean _in general_. I'm not advising against writing tests. But in this case, I don't see any benefit in testing that a function returning a string actually returns the string—that sounds more like a constant, and I don't think we test constants, do we 🙂.

File: packages/web/src/scss/components/Modal/index.html (hunk 5 above)
Comment (adamkudrna) [reply packages/web/src/scss/components/Modal/index.html:792]: Eventually, I changed it right away, rebase is still quite safe.
